### ISSUE TITLE
Eliminate TileGrad in favor of reshape followed by reduce_sum

### ISSUE THREAD

**Issue by girving** [tensorflow/tensorflow#643]
`TileGrad` is exactly equivalent to a reshape which splits each dimension into two pieces, an input dimension piece and a tile dimension piece, then does a reduce_sum over all the tile dimensions to get back the input shape.  There is no need for a separate op.

For example, if `x.shape = (2, 3)` and we do `tf.tile(x, [5, 7])`, the tiled tensor has shape `(2 * 5, 3 * 7)`.  Given a gradient `y` w.r.t. that output with shape `(2 * 5, 3 * 7)`, we can recover the input gradient as `tf.reduce_sum(tf.reshape(y, [5, 2, 7, 3]), reduction_indices=[0, 2])`.

Note that this works even in the case where `multiples` has zeros: in that case we'll do a `reduce_sum` where some of the dimensions we sum over will be zero.  This is well defined even if the result after summation is nonempty: in that case the result is zero.  If sum doesn't already work in that case, it should.

**Comment by girving** [comment]
@benoitsteiner: Is there a different reason why a separate `TileGrad` op is required?

**Comment by girving** [comment]
I fixed this a while ago.
